Math/NumberTheory/Moduli/CubicSymbol.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Math.NumberTheory.Moduli.CubicSymbol
+    ( CubicSymbol(..)
+    , conj
+    , cubicSymbol
+    ) where
+
+import qualified Math.NumberTheory.Quadratic.EisensteinIntegers as E
+import qualified Math.NumberTheory.Primes as P
+import qualified Data.Euclidean as A
+import qualified Data.Semigroup as S
+import qualified Data.List as L
+
+
+
+data CubicSymbol = Zero | Omega | OmegaSquare | One deriving (Eq)
+
+instance S.Semigroup CubicSymbol where
+    (<>) = multiplication
+
+instance Show CubicSymbol where
+    show = \case
+        Zero         -> show 0
+        Omega        -> "ω"
+        OmegaSquare  -> "ω²"
+        One          -> show 1
+
+
+multiplication :: CubicSymbol -> CubicSymbol -> CubicSymbol
+multiplication Zero _ = Zero
+multiplication _ Zero = Zero
+multiplication One y = y
+multiplication x One = x
+multiplication Omega Omega = OmegaSquare
+multiplication Omega OmegaSquare = One
+multiplication OmegaSquare Omega = One
+multiplication OmegaSquare OmegaSquare = Omega
+
+
+
+exponentiation :: CubicSymbol -> Integer -> CubicSymbol
+exponentiation Zero _ = Zero
+exponentiation One _ = One
+exponentiation _ 0 = One
+exponentiation Omega 1 = Omega
+exponentiation Omega 2 = OmegaSquare
+exponentiation OmegaSquare 1 = OmegaSquare
+exponentiation OmegaSquare 2 = Omega
+
+
+
+conj :: CubicSymbol -> CubicSymbol
+conj = \case
+    Zero          -> Zero
+    Omega         -> OmegaSquare
+    OmegaSquare   -> Omega
+    One           -> One
+
+-- Converts Eisenstein integer to CubicSymbol value.
+toCubicSymbol :: E.EisensteinInteger -> CubicSymbol
+toCubicSymbol = \case
+    0               -> Zero
+    0 E.:+ 1        -> Omega
+    (-1) E.:+ (-1)  -> OmegaSquare
+    1               -> One
```

Comment:
I am not sure I understand this issue. I added the exhaustive pattern matching and a not very meaningful error message. However the compiler gives me a warning saying that "otherwise" is redundant.